server/sonar-server/src/main/java/org/sonar/server/computation/step/FeedPeriodsStep.java

```diff
@@ -0,0 +1,279 @@
+      long targetDate = DateUtils.addDays(new Date(analysisDate), -days).getTime();
+      SnapshotDto snapshot = findNearestSnapshotToTargetDate(snapshots, targetDate);
+      if (snapshot == null) {
+        return null;
+      }
+      LOG.debug(String.format("Compare over %s days (%s, analysis of %s)", String.valueOf(days), formatDate(targetDate), formatDate(snapshot.getCreatedAt())));
+      return new Period(index, CoreProperties.TIMEMACHINE_MODE_DAYS, String.valueOf(days), snapshot.getCreatedAt());
+    }
+
+    @CheckForNull
+    private Period findByPreviousAnalysis(int index) {
+      SnapshotDto snapshot = findFirstSnapshot(session, createCommonQuery(projectId).setCreatedBefore(analysisDate).setIsLast(true).setSort(BY_DATE, DESC));
+      if (snapshot == null) {
+        return null;
+      }
+      LOG.debug(String.format("Compare to previous analysis (%s)", formatDate(snapshot.getCreatedAt())));
+      return new Period(index, CoreProperties.TIMEMACHINE_MODE_PREVIOUS_ANALYSIS, formatDate(snapshot.getCreatedAt()), snapshot.getCreatedAt());
+    }
+
+    @CheckForNull
+    private Period findByPreviousVersion(int index) {
+      List<SnapshotDto> snapshotDtos = dbClient.snapshotDao().selectPreviousVersionSnapshots(session, projectId, currentVersion);
+      if (snapshotDtos.isEmpty()) {
+        return null;
+      }
+      SnapshotDto snapshotDto = snapshotDtos.get(0);
+      LOG.debug(String.format("Compare to previous version (%s)", formatDate(snapshotDto.getCreatedAt())));
+      return new Period(index, CoreProperties.TIMEMACHINE_MODE_PREVIOUS_VERSION, snapshotDto.getVersion(), snapshotDto.getCreatedAt());
+    }
+
+    @CheckForNull
+    private Period findByVersion(int index, String version) {
+      SnapshotDto snapshot = findFirstSnapshot(session, createCommonQuery(projectId).setVersion(version).setSort(BY_DATE, DESC));
+      if (snapshot == null) {
+        return null;
+      }
+      LOG.debug(String.format("Compare to version (%s) (%s)", version, formatDate(snapshot.getCreatedAt())));
+      return new Period(index, CoreProperties.TIMEMACHINE_MODE_VERSION, version, snapshot.getCreatedAt());
+    }
+
+    @CheckForNull
+    private SnapshotDto findFirstSnapshot(DbSession session, SnapshotQuery query) {
+      List<SnapshotDto> snapshots = dbClient.snapshotDao().selectSnapshotsByQuery(session, query);
+      if (!snapshots.isEmpty()) {
+        return snapshots.get(0);
+      }
+      return null;
+    }
+  }
+
+  @CheckForNull
+  private static Integer tryToResolveByDays(String property) {
+    try {
+      return Integer.parseInt(property);
+    } catch (NumberFormatException e) {
+      return null;
+    }
+  }
+
+  @CheckForNull
+  private static Date tryToResolveByDate(String property) {
+    try {
+      return DateUtils.parseDate(property);
+    } catch (Exception e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)